File: lib/src/main/java/password/manager/lib/ReadablePasswordFieldWithStrSkin.java
/*
    Password Manager: Manages accounts given by user with encrypted password.
    Copyright (C) 2022-2026  Francesco Marras (2004marras@gmail.com)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see https://www.gnu.org/licenses/gpl-3.0.html.
 */

package password.manager.lib;

import static password.manager.lib.Utils.*;

import java.util.function.Function;

import javafx.animation.KeyValue;
import javafx.scene.Node;
import javafx.scene.control.ProgressBar;
import javafx.scene.layout.Background;

public class ReadablePasswordFieldWithStrSkin extends ReadablePasswordFieldSkin implements AnimationAwareControl {

    private static final Function<String, Double> PROGRESS_EXTRACTOR = pass -> {
        double passwordStrength = passwordStrength(pass);
        return doubleSquash(0d, (passwordStrength - 20d) / 30d, 1d);
    };

    private final ProgressBar strengthBar;
    private final AnimationController<String> animationController;

    public ReadablePasswordFieldWithStrSkin(CustomPasswordField control) {
        this(control, false);
    }

    public ReadablePasswordFieldWithStrSkin(CustomPasswordField control, boolean initialReadable) {
        super(control, initialReadable);

        this.strengthBar = new ProgressBar(0);

        strengthBar.setManaged(false);
        strengthBar.setFocusTraversable(false);
        strengthBar.setPickOnBounds(false);
        
        strengthBar.setVisible(true);
        strengthBar.setBackground(Background.EMPTY);

        strengthBar.toFront();
        getChildren().add(strengthBar);

        final Function<Double, KeyValue> STYLE_FUNCTION = prog -> {
            Node bar = strengthBar.lookup(".bar");
            if(bar == null) return null;
            return new KeyValue(bar.styleProperty(), "-fx-background-color:" + passwordStrengthGradient(prog) + ";");
        };

        this.animationController = new AnimationController<>(
            control.textProperty(),
            PROGRESS_EXTRACTOR,
            strengthBar.progressProperty(),
            STYLE_FUNCTION,
            strengthBar
        );
    }

    @Override
    protected void layoutChildren(double x, double y, double w, double h) {
        super.layoutChildren(x, y, w, h);

        double controlW = getSkinnable().getWidth();
        double controlH = getSkinnable().getHeight();

        double barH = 10;
        double spacing = barH / 4;
        strengthBar.resizeRelocate(spacing / 2, y + controlH + 2 * spacing - barH, controlW - spacing, barH);
    }

    @Override
    public void dispose() {
        animationController.detach();
        super.dispose();
    }

    ///// ANIMATION AWARE CONTROL METHODS /////

    @Override
    public AnimationController<?> getAnimationController() {
        return animationController;
    }

    @Override
    public void onListenerDetached() {
        strengthBar.setVisible(false);
    }

    @Override
    public void onListenerAttached() {
        strengthBar.setVisible(true);
    }
}

File: lib/src/main/kotlin/password/manager/lib/ReadablePasswordFieldWithStrSkin.kt
/*
    Password Manager: Manages accounts given by user with encrypted password.
    Copyright (C) 2022-2026  Francesco Marras (2004marras@gmail.com)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see https://www.gnu.org/licenses/gpl-3.0.html.
 */
package password.manager.lib

import javafx.animation.KeyValue
import javafx.scene.Node
import javafx.scene.control.ProgressBar
import javafx.scene.layout.Background
import password.manager.lib.Utils.passwordStrength
import password.manager.lib.Utils.passwordStrengthGradient

class ReadablePasswordFieldWithStrSkin @JvmOverloads constructor(
    control: CustomPasswordField,
    initialReadable: Boolean = false
) : ReadablePasswordFieldSkin(control, initialReadable), AnimationAwareControl {
    private val strengthBar: ProgressBar = ProgressBar(0.0)
    private val animContr: AnimationController<String>

    init {
        strengthBar.isManaged = false
        strengthBar.isFocusTraversable = false
        strengthBar.isPickOnBounds = false

        strengthBar.isVisible = true
        strengthBar.background = Background.EMPTY

        strengthBar.toFront()
        children.add(strengthBar)

        val styleFunction: (Double) -> KeyValue? = { prog: Double ->
            val bar: Node? = strengthBar.lookup(".bar")
            bar?.let { KeyValue(it.styleProperty(), "-fx-background-color:${passwordStrengthGradient(prog)};") }
        }

        animContr = AnimationController<String>(
            control.textProperty(),
            PROGRESS_EXTRACTOR,
            strengthBar.progressProperty(),
            styleFunction,
            strengthBar
        )
    }

    @Suppress("UNNECESSARY_SAFE_CALL")
    override fun layoutChildren(x: Double, y: Double, w: Double, h: Double) {
        super.layoutChildren(x, y, w, h)

        val controlW = skinnable.width
        val controlH = skinnable.height

        val barH = 10.0
        val spacing = barH / 4
        strengthBar?.resizeRelocate(spacing / 2, y + controlH + 2 * spacing - barH, controlW - spacing, barH)
    }

    @Suppress("UNNECESSARY_SAFE_CALL")
    override fun dispose() {
        animationController?.detach()
        super.dispose()
    }

    ///// ANIMATION AWARE CONTROL METHODS //////
    override val animationController: AnimationController<*>
        get() = animContr

    override fun onListenerDetached() {
        strengthBar.isVisible = false
    }

    override fun onListenerAttached() {
        strengthBar.isVisible = true
    }

    companion object {
        private val PROGRESS_EXTRACTOR: (String) -> Double = { pass: String ->
            val passStr = passwordStrength(pass)
            ((passStr.first - 20.0) / 30.0).coerceIn(0.0, 1.0)
        }
    }
}
